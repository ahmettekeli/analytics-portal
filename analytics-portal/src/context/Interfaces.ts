import { actionTypes } from "./ActionTypes";

//base interface olusturup app ve campaign icin onlardan turetilebilir.
//2 dosya olsun, app.interface.ts ve campaign.interface.ts
//app.types.ts ve campaign.types.ts olabilir.

//Install interface {
//   day:
//   value:
//   id?:
//   appId?:
// }

export interface AppInstallInterface {
  day: string;
  value: number;
  id: string;
  appId: string;
}

export interface RevenueInterface {
  day: string;
  value: number;
  id: string;
  appId: string;
}

export interface CampaignInstallInterface {
  day: string;
  value: number;
}

export interface AppDataInterface {
  createdAt: Date;
  name: string;
  icon: string;
  active: boolean;
  id: string;
  installs: AppInstallInterface[];
  revenue: RevenueInterface[];
  campaigns: CampaignType[];
}

export type CampaignType = {
  name: string;
  installs: CampaignInstallInterface[];
  id: string;
  appId?: string;
};

export type AnalyticsStateType =
  | {
      appData: AppDataInterface[];
      currentApp: AppDataInterface;
      isLoading: boolean;
      errorMessage: string;
    }
  | {
      appData: AppDataInterface[];
      currentApp: AppDataInterface | undefined;
      isLoading: boolean;
      errorMessage: null;
    };

export type AppCampaignType = {
  app: AppDataInterface;
  campaign: CampaignType;
};

export type ContextActionType =
  | { type: actionTypes.GET_ANALYTICS_DATA; payload: AnalyticsStateType }
  | {
      type: actionTypes.ADD_CAMPAIGN;
      payload: AppCampaignType;
    }
  | {
      type: actionTypes.SET_CURRENT_APP;
      payload: AppDataInterface;
    };

export type AppChartDataType = {
  installLabels: string[];
  installData: number[];
  revenueLabels: string[];
  revenueData: number[];
};

export type CampaignChartDataType = {
  installLabels: string[];
  installData: number[];
};
