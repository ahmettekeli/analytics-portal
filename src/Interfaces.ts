import { actionTypes } from "./context/ActionTypes";
import { CampaignType } from "campaign.types";

export interface InstallInterface {
  day: string;
  value: number;
  id?: string;
  appId?: string;
}

export interface AppInstallInterface extends InstallInterface {}

export interface RevenueInterface {
  day: string;
  value: number;
  id: string;
  appId: string;
}

export interface CampaignInstallInterface extends InstallInterface {}

export interface AppDataInterface {
  createdAt: Date;
  name: string;
  icon: string;
  active: boolean;
  id: string;
  installs: AppInstallInterface[];
  revenue: RevenueInterface[];
  campaigns: CampaignType[];
}

export type AnalyticsStateType =
  | {
      appData: AppDataInterface[];
      currentApp: AppDataInterface;
      isLoading: boolean;
      errorMessage: string;
    }
  | {
      appData: AppDataInterface[];
      currentApp: AppDataInterface | undefined;
      isLoading: boolean;
      errorMessage: null;
    };

export type AppCampaignType = {
  app: AppDataInterface;
  campaign: CampaignType;
};

export type ContextActionType =
  | { type: actionTypes.GET_ANALYTICS_DATA; payload: AnalyticsStateType }
  | {
      type: actionTypes.ADD_CAMPAIGN;
      payload: AppCampaignType;
    }
  | {
      type: actionTypes.SET_CURRENT_APP;
      payload: AppDataInterface;
    };
