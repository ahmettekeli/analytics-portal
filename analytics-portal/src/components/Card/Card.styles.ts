import styled from "styled-components";
import { Link } from "react-router-dom";
import { colors } from "configs";

export const Wrapper = styled.div`
  display: flex;
  justify-content: space-between;
  background-color: ${colors.cardBackground};
  margin: 1rem 0;
  border-radius: 0.5rem;
  width: 100%;

  @media (max-width: 350px) {
    width: 350px;
  }
`;

export const StyledLink = styled(Link)`
  display: flex;
  text-decoration: none;
  width: 100%;
  justify-content: space-between;
  &:focus,
  &:hover,
  &:visited,
  &:link,
  &:active {
    text-decoration: none;
  }
`;
