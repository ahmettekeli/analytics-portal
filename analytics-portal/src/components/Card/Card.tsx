import {
  CardDetailElement,
  CardProfile,
  CardProfileImg,
  CardDetail,
  NameContainer,
  NumberContainer,
  Wrapper,
  StyledLink,
} from "./Card.styles";
import capitalize from "lodash/capitalize";
import { actionTypes } from "../../context/ActionTypes";
import { AppDataInterface } from "../../context/Interfaces";
import { getAvgInstalls, getAvgRevenue } from "../../utils/utilities";
import { useAPI } from "../../context/Store";

function Card({ app }: { app: AppDataInterface }) {
  const { dispatch } = useAPI();
  const { active, campaigns, icon, name } = app;
  const avgRevenue = getAvgRevenue(app, campaigns.length);
  const avgInstalls = getAvgInstalls(app, campaigns.length);

  //TODO currentApp doesn't get set in the global store if you enter a url /overview/:name
  //find a way to dispatch with react router
  return (
    <Wrapper data-testid="Card">
      <CardProfile isActive={active}>
        <StyledLink
          to={`/overview/${name}`}
          onClick={() => {
            dispatch({ type: actionTypes.SET_CURRENT_APP, payload: app });
          }}
        >
          <CardProfileImg src={icon} alt={name} />
        </StyledLink>
        <StyledLink
          to={`/overview/${name}`}
          onClick={() => {
            dispatch({ type: actionTypes.SET_CURRENT_APP, payload: app });
          }}
        >
          <NameContainer>{capitalize(name)}</NameContainer>
        </StyledLink>
      </CardProfile>
      <CardDetail isActive={active}>
        <CardDetailElement>
          <NumberContainer>{campaigns.length}</NumberContainer>
          <p>Campaigns</p>
        </CardDetailElement>
        <CardDetailElement>
          <NumberContainer>{avgInstalls}</NumberContainer>
          <p>Avg. Installs</p>
        </CardDetailElement>
        <CardDetailElement>
          <NumberContainer>{avgRevenue}</NumberContainer>
          <p>Avg. Revenue</p>
        </CardDetailElement>
      </CardDetail>
    </Wrapper>
  );
}
export default Card;
