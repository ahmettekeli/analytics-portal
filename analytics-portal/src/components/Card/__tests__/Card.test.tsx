import { cleanup, render } from "@testing-library/react";
import renderer from "react-test-renderer";
import Card from "../Card";
import { BrowserRouter as Router } from "react-router-dom";
import { AppDataInterface } from "Interfaces";

describe("Card", () => {
  let renderResult: ReturnType<typeof render>;
  let activeApp: AppDataInterface = {
    createdAt: new Date(),
    name: "Active test app",
    icon: "http://placeimg.com/640/480/abstract",
    active: true,
    id: "1",
    installs: [
      {
        day: "day 1",
        value: 61,
        id: "1",
        appId: "1",
      },
      {
        day: "day 6",
        value: 54,
        id: "6",
        appId: "1",
      },
      {
        day: "day 11",
        value: 88,
        id: "11",
        appId: "1",
      },
      {
        day: "day 16",
        value: 16,
        id: "16",
        appId: "1",
      },
      {
        day: "day 21",
        value: 48,
        id: "21",
        appId: "1",
      },
    ],
    revenue: [
      {
        day: "day 1",
        value: 88,
        id: "1",
        appId: "1",
      },
      {
        day: "day 6",
        value: 79,
        id: "6",
        appId: "1",
      },
      {
        day: "day 11",
        value: 59,
        id: "11",
        appId: "1",
      },
      {
        day: "day 16",
        value: 45,
        id: "16",
        appId: "1",
      },
      {
        day: "day 21",
        value: 16,
        id: "21",
        appId: "1",
      },
    ],
    campaigns: [
      {
        name: "Charlene Kohler",
        installs: [
          {
            day: "Day 1",
            value: 35,
          },
          {
            day: "Day 2",
            value: 15,
          },
          {
            day: "Day 3",
            value: 2,
          },
          {
            day: "Day 4",
            value: 0,
          },
          {
            day: "Day 5",
            value: 1,
          },
          {
            day: "Day 6",
            value: 1,
          },
          {
            day: "Day 7",
            value: 0,
          },
        ],
        id: "1",
      },
    ],
  };
  beforeEach(() => {
    renderResult = render(
      <Router>
        <Card app={activeApp} />
      </Router>
    );
  });
  afterEach(cleanup);

  test("Renders Card component", () => {
    expect(renderResult.getByTestId("card")).toBeInTheDocument();
  });

  //TODO
  test("Renders inactive Card faded", () => {});

  test("Card has correct image and name", () => {
    expect(
      renderResult.getByTestId("card").getElementsByTagName("img")[0].src
    ).toBe(activeApp.icon);
    expect(renderResult.getByText(activeApp.name)).toBeInTheDocument();
  });

  test("Card has correct campaigns value", () => {});
  //TODO
  test("Card has correct avg installs value", () => {});
  //TODO
  test("Card has correct avg revenue value", () => {});
  //TODO

  test("Card matches snapshot", () => {
    const tree = renderer
      .create(
        <Router>
          <Card app={activeApp} />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
