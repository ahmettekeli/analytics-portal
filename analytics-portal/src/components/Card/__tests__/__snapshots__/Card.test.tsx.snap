// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Card matches snapshot 1`] = `
<div
  className="sc-bdvvtL eSfVxr"
  data-testid="Card"
>
  <div
    className="sc-gsDKAQ geNvXo"
  >
    <a
      className="sc-iCfMLu iMJIiJ"
      href="/overview/Active test app"
      onClick={[Function]}
    >
      <img
        alt="Active test app"
        className="sc-dkPtRN hBlbtm"
        src="http://placeimg.com/640/480/abstract"
      />
    </a>
    <a
      className="sc-iCfMLu iMJIiJ"
      href="/overview/Active test app"
      onClick={[Function]}
    >
      <div
        className="sc-jRQBWg inHIzc"
      >
        Active test app
      </div>
    </a>
  </div>
  <div
    className="sc-hKwDye iHlHRv"
  >
    <div
      className="sc-eCImPb chOnSp"
    >
      <p
        className="sc-gKclnd cRtMhS"
      >
        1
      </p>
      <p>
        Campaigns
      </p>
    </div>
    <div
      className="sc-eCImPb chOnSp"
    >
      <p
        className="sc-gKclnd cRtMhS"
      >
        267
      </p>
      <p>
        Avg. Installs
      </p>
    </div>
    <div
      className="sc-eCImPb chOnSp"
    >
      <p
        className="sc-gKclnd cRtMhS"
      >
        287
      </p>
      <p>
        Avg. Revenue
      </p>
    </div>
  </div>
</div>
`;
